macros/edr/data_monitoring/monitors_query/table_monitoring_query.sql

```diff
@@ -144,4 +91,183 @@
+        {{ metric_name }} as (
+            {{ metric_query }}
+        ){% if not loop.last %},{% endif %}
+    {%- endfor %}
+
+    {%- for metric_name in included_monitors %}
+    select * from {{ metric_name }}
+    {% if not loop.last %} union all {% endif %}
+    {%- endfor %}
+{% endmacro %}
+
+{% macro get_metric_query(metric_name, metric_args, timestamp_column) %}
+    {%- set metrics_macro_mapping = {
+        "row_count": elementary.row_count_metric_query,
+        "freshness": elementary.freshness_metric_query,
+        "event_freshness": elementary.event_freshness_metric_query
+    } %}
+
+    {%- set metric_macro = metrics_macro_mapping.get(metric_name) %}
+    {%- if not metric_macro %}
+        {%- do return(none) %}
+    {%- endif %}
+
+    {%- set metric_query = metric_macro(metric_args, timestamp_column=timestamp_column) %}
+    {%- if not metric_query %}
+        {%- do return(none) %}
+    {%- endif %}
+
+    {{ metric_query }}
+{% endmacro %}
+
+{% macro row_count_metric_query(metric_args, timestamp_column=none) %}
+{% if timestamp_column %}
+    with row_count_values as (
+        select edr_bucket_start,
+               edr_bucket_end,
+               start_bucket_in_data,
+               case when start_bucket_in_data is null then
+                   0
+               else {{ elementary.cast_as_float(elementary.row_count()) }} end as row_count_value
+        from buckets left join time_filtered_monitored_table on (edr_bucket_start = start_bucket_in_data)
+        group by 1,2,3
+    )
+
+    select edr_bucket_start,
+           edr_bucket_end,
+           {{ elementary.const_as_string('row_count') }} as metric_name,
+           {{ elementary.null_string() }} as source_value,
+           row_count_value as metric_value
+    from row_count_values
+{% else %}
+    select
+        {{ elementary.const_as_string('row_count') }} as metric_name,
+        {{ elementary.row_count() }} as metric_value
+    from monitored_table
+    group by 1
+{% endif %}
+{% endmacro %}
+
+{% macro freshness_metric_query(metric_args, timestamp_column=none) %}
+{% if timestamp_column %}
+    {%- set freshness_column = metric_args.freshness_column %}
+    {%- if freshness_column is undefined or freshness_column is none %}
+        {%- set freshness_column = timestamp_column %}
+    {%- endif %}
+
+    -- get ordered consecutive update timestamps in the source data
+    with unique_timestamps as (
+        select distinct {{ elementary.cast_as_timestamp(freshness_column) }} as timestamp_val
+        from monitored_table
+        order by 1
+    ),
+
+    -- compute freshness for every update as the time difference from the previous update
+    consecutive_updates_freshness as (
+        select
+            timestamp_val as update_timestamp,
+            lag(timestamp_val) over (order by timestamp_val) as prev_timestamp,
+            {{ elementary.timediff('second', 'prev_timestamp', 'update_timestamp') }} as freshness
```

Comment:
This command doesn't works in BQ.
I get an error `Unrecognized name: prev_timestamp at [121:20]`
Seems like you can't reference to field that defined in the same select.
changing the query to the following solved the issue
`timestamp_diff(lag(timestamp_val) over (order by timestamp_val), timestamp_val, second)`